core/client/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -540,11 +544,26 @@ private void updateCacheStream(long blockId) throws IOException {
     }
 
     try {
-      WorkerNetAddress address = mLocationPolicy.getWorkerForNextBlock(
-          mBlockStore.getWorkerInfoList(), getBlockSizeAllocation(mPos));
-      // If we reach here, we need to cache.
-      mCurrentCacheStream =
-          mBlockStore.getOutStream(blockId, getBlockSize(mPos), address, mOutStreamOptions);
+      List<BlockWorkerInfo> workers = mBlockStore.getWorkerInfoList();
```

Comment:
I've removed the write lock from the worker side since temp blocks should be private to a single session anyway. 

It's difficult on the client side to know how to handle `BlockAlreadyExistException` because different implementations of the `BlockInStream` might have different behaviors around when they realize that the block already exists and under what circumstances they need to be canceled/closed. I think it's possible to define contracts to clean this up, but it's a nontrivial amount of work. 

1. Changed to only cache to local when reading from remote.
2. I don't think `getOutStream` has the worker list information, but I've changed `mContext` to hold a local worker address which can be used to avoid calling `getWorkerInfoIist`.